### ISSUE TITLE
Experiment with various galera params

### ISSUE THREAD

**Issue by sanikolaev** [manticoresoftware/manticoresearch-buddy#255]
In https://github.com/manticoresoftware/manticoresearch-buddy/commit/a4671310b938a09df9ec415903596e72ba25c6d7#diff-80834d764c22d490e12c72a69b39bfc8a03f87ec042903e3f14938668d8d74b2R13-R15 we have a few galera params that supposedly help `create table shards=N ...` not to fail when the number of shards is high (like 60). Let's experiment with the params to figure out which one helps most. Chances are it works without them at all now.

**Comment by donhardman** [comment]
I tested it without changing any options, and I encountered the same issue. After narrowing it down, I found that only one config setting really affects it and helps avoid errors in our test. That setting is `gmcast.peer_timeout=PT5S`.

I left only that setting and pushed it to the main branch for now, but it's probably a good idea to consider tuning it, maybe on the Manticore daemon side. 

The default timeout in Galera is PT3S. So increasing this option helped to stabilize sharding. 

I scored 30 runs without any errors, compared to only 9 runs before getting an error without the change.

**Comment by sanikolaev** [comment]
> That setting is gmcast.peer_timeout=PT5S

Great! As discussed, pls try to reproduce the issue with a smaller timeout and a simpler case. We may want to make some improvements in the code around "error in node state, must reinit", so the timeout is not that fatal.

**Comment by donhardman** [comment]
It seems like the issue is fully dependent on the network/server load. I've tried to find a Minimal Reproducible Example to reproduce it, but I didn't succeed. After that, I tried running the same test with the PT1S config that reduces the timeout, and all 50 runs succeeded. So I think the issue is highly dependent on server load, and it's very problematic to reproduce even though it occurs rarely. I'm not sure what exactly triggers it.

**Comment by sanikolaev** [comment]
As discussed, pls try `PT0.1S`

**Comment by donhardman** [comment]
Afterter setting perr timeout to PT0.1S the test  fails everytime we run it. Here is the log with replication log enabled.

[create-table-5-cluster-60-shards.rep.zip](https://github.com/manticoresoftware/manticoresearch-buddy/files/15009320/create-table-5-cluster-60-shards.rep.zip)

**Comment by tomatolog** [comment]
seems issue with Galera - I'd like to build and try vanila Galera library or use one `galera.so` from one of the mysql distribution instead of the Percona fork we use.

However to do so I need to replicate issue locally at the dev2 box.
I tried to use buddy master version 73ec2463aeefeb5c057a6a2c4cd5025b32e90e25 and there buddy fails to create sharding table
```
mysql> create table c:t(id bigint) shards=60 rf=1;
ERROR 1064 (42000): Waiting timeout exceeded.
```
however at the daemon log I do not see the any `create cluster` statements or error message `Queue query result [error]: {"error":"cluster 'e821b66c530a4aec4a29fec79fd45a4f' is not ready, not primary state (synced)"}`

I also tried to use buddy from the branch `sharding/inserts` a89df531ffd20a20d0db5a34d2d06339ba0083b9 however daemon failed to start buddy of that version with the error message
```
[Thu Apr 25 13:05:46.540 2024] [3278333] [BUDDY] Loaded plugins:
[Thu Apr 25 13:05:46.540 2024] [3278333] [BUDDY]   core: empty-string, backup, emulate-elastic, distributed-insert, insert, alias, select, show, cli-ta
[Thu Apr 25 13:05:46.540 2024] [3278333] [BUDDY]   local:
[Thu Apr 25 13:05:46.540 2024] [3278333] [BUDDY]   extra:
[Thu Apr 25 13:05:46.540 2024] [3278333] [BUDDY] Error while starting the server: Node ID detection failed
[Thu Apr 25 13:05:46.554 2024] [3278334] WARNING: [BUDDY] terminated, exit code 0
```

I need a way to reproduce this case locally to investigate the issue further

**Comment by donhardman** [comment]
I've rebased the `sharding/inserts` branch from the latest main, and it looks like it's working now. 👍

We should use the `sharding/inserts` branch because it's the only one where we've set the timeout to PT0.1S, while in the main branch, it's set to PT5S, which won't allow us to debug the issue effectively.

In configurations 2 and 3, we have double colons `::`. If we replace them with a single colon `:`, it should help solve the issue with failed node detection.

```
[BUDDY]   string(16) "127.0.0.1:31212"
```

![Image](https://github.com/manticoresoftware/manticoresearch-buddy/assets/59067742/ece50569-041a-4851-a573-156ae133fcc3)

**Comment by tomatolog** [comment]
I unable to reproduce the case locally at the dev box
I use 5 nodes at the `~/bin/rpl` and every time buddy returns error
```
mysql> create table c:a(id bigint) shards=60 rf=3;
ERROR 1064 (42000): Waiting timeout exceeded.
```

but looking at the daemon I see the clusters are healthy
```
mysql> show status like '%cluster%stat%';
+--------------------------------------------------------------+--------------------------------------+
| Counter                                                      | Value                                |
+--------------------------------------------------------------+--------------------------------------+
| cluster_c_state_uuid                                         | e0fea276-0633-11ef-88d5-fb3a6b6fb060 |
| cluster_c_status                                             | primary                              |
| cluster_c_node_state                                         | synced                               |
| cluster_c_local_state_uuid                                   | e0fea276-0633-11ef-88d5-fb3a6b6fb060 |
| cluster_c_flow_control_status                                | OFF                                  |
| cluster_c_local_state                                        | 4                                    |
| cluster_c_local_state_comment                                | Synced                               |
| cluster_c_ist_receive_status                                 |                                      |
| cluster_c_evs_state                                          | OPERATIONAL                          |
| cluster_c63021ddef800c5c809e4855b75bd80b_state_uuid          | f2e1017e-0633-11ef-a62c-3bd47a66d612 |
| cluster_c63021ddef800c5c809e4855b75bd80b_status              | primary                              |
| cluster_c63021ddef800c5c809e4855b75bd80b_node_state          | synced                               |
| cluster_c63021ddef800c5c809e4855b75bd80b_local_state_uuid    | f2e1017e-0633-11ef-a62c-3bd47a66d612 |
| cluster_c63021ddef800c5c809e4855b75bd80b_flow_control_status | OFF                                  |
| cluster_c63021ddef800c5c809e4855b75bd80b_local_state         | 4                                    |
| cluster_c63021ddef800c5c809e4855b75bd80b_local_state_comment | Synced                               |
| cluster_c63021ddef800c5c809e4855b75bd80b_ist_receive_status  |                                      |
| cluster_c63021ddef800c5c809e4855b75bd80b_evs_state           | OPERATIONAL                          |
| cluster_d3bc8aab1dfe8412f9f00052865b482a_state_uuid          | f31c9b78-0633-11ef-ad50-469d13cb98c4 |
| cluster_d3bc8aab1dfe8412f9f00052865b482a_status              | primary                              |
| cluster_d3bc8aab1dfe8412f9f00052865b482a_node_state          | synced                               |
| cluster_d3bc8aab1dfe8412f9f00052865b482a_local_state_uuid    | f31c9b78-0633-11ef-ad50-469d13cb98c4 |
| cluster_d3bc8aab1dfe8412f9f00052865b482a_flow_control_status | OFF                                  |
| cluster_d3bc8aab1dfe8412f9f00052865b482a_local_state         | 4                                    |
| cluster_d3bc8aab1dfe8412f9f00052865b482a_local_state_comment | Synced                               |
| cluster_d3bc8aab1dfe8412f9f00052865b482a_ist_receive_status  |                                      |
| cluster_d3bc8aab1dfe8412f9f00052865b482a_evs_state           | OPERATIONAL                          |
| cluster_d40cdb7381e229c88c1dec8e8df0ae20_state_uuid          | f35ea437-0633-11ef-94cd-9b5ef3028b24 |
| cluster_d40cdb7381e229c88c1dec8e8df0ae20_status              | primary                              |
| cluster_d40cdb7381e229c88c1dec8e8df0ae20_node_state          | synced                               |
| cluster_d40cdb7381e229c88c1dec8e8df0ae20_local_state_uuid    | f35ea437-0633-11ef-94cd-9b5ef3028b24 |
| cluster_d40cdb7381e229c88c1dec8e8df0ae20_flow_control_status | OFF                                  |
| cluster_d40cdb7381e229c88c1dec8e8df0ae20_local_state         | 4                                    |
| cluster_d40cdb7381e229c88c1dec8e8df0ae20_local_state_comment | Synced                               |
| cluster_d40cdb7381e229c88c1dec8e8df0ae20_ist_receive_status  |                                      |
| cluster_d40cdb7381e229c88c1dec8e8df0ae20_evs_state           | OPERATIONAL                          |
| cluster_d8783a3642045d36f66ec12ff41b402c_state_uuid          | f3a18cf7-0633-11ef-964e-16f732a531ef |
| cluster_d8783a3642045d36f66ec12ff41b402c_status              | primary                              |
| cluster_d8783a3642045d36f66ec12ff41b402c_node_state          | synced                               |
| cluster_d8783a3642045d36f66ec12ff41b402c_local_state_uuid    | f3a18cf7-0633-11ef-964e-16f732a531ef |
| cluster_d8783a3642045d36f66ec12ff41b402c_flow_control_status | OFF                                  |
| cluster_d8783a3642045d36f66ec12ff41b402c_local_state         | 4                                    |
| cluster_d8783a3642045d36f66ec12ff41b402c_local_state_comment | Synced                               |
| cluster_d8783a3642045d36f66ec12ff41b402c_ist_receive_status  |                                      |
| cluster_d8783a3642045d36f66ec12ff41b402c_evs_state           | OPERATIONAL                          |
| cluster_af5f3c1561e6c967c6337599461cf246_state_uuid          | f3ff04e6-0633-11ef-9336-677447aa69a5 |
| cluster_af5f3c1561e6c967c6337599461cf246_status              | primary                              |
| cluster_af5f3c1561e6c967c6337599461cf246_node_state          | synced                               |
| cluster_af5f3c1561e6c967c6337599461cf246_local_state_uuid    | f3ff04e6-0633-11ef-9336-677447aa69a5 |
| cluster_af5f3c1561e6c967c6337599461cf246_flow_control_status | OFF                                  |
| cluster_af5f3c1561e6c967c6337599461cf246_local_state         | 4                                    |
| cluster_af5f3c1561e6c967c6337599461cf246_local_state_comment | Synced                               |
| cluster_af5f3c1561e6c967c6337599461cf246_ist_receive_status  |                                      |
| cluster_af5f3c1561e6c967c6337599461cf246_evs_state           | OPERATIONAL                          |
| cluster_ecb841b63a92fbe66097296d2d1c8375_state_uuid          | fa3adfe4-0633-11ef-a1f0-73ad06525620 |
| cluster_ecb841b63a92fbe66097296d2d1c8375_status              | primary                              |
| cluster_ecb841b63a92fbe66097296d2d1c8375_node_state          | synced                               |
| cluster_ecb841b63a92fbe66097296d2d1c8375_local_state_uuid    | fa3adfe4-0633-11ef-a1f0-73ad06525620 |
| cluster_ecb841b63a92fbe66097296d2d1c8375_flow_control_status | OFF                                  |
| cluster_ecb841b63a92fbe66097296d2d1c8375_local_state         | 4                                    |
| cluster_ecb841b63a92fbe66097296d2d1c8375_local_state_comment | Synced                               |
| cluster_ecb841b63a92fbe66097296d2d1c8375_ist_receive_status  |                                      |
| cluster_ecb841b63a92fbe66097296d2d1c8375_evs_state           | OPERATIONAL                          |
+--------------------------------------------------------------+--------------------------------------+
63 rows in set (0.00 sec)
```

daemon logs have no error `is not ready, not primary state (synced)` it is not clear why buddy reply with the error message and what cause it.

Need someway to make the case clear and collect some commands these cause buddy to breaks.

**Comment by donhardman** [comment]
I can confirm that it's not reproducible.

So I have lowered the timeout in sources from PT0.1S to PT0.05S because it looks like the network is stable, and after that, I found some weird issues in the log:

```
[Tue Apr 30 05:20:39.406 2024] [312651] DEBUG: P01: syntax error, unexpected identifier near
cluster_c_evs_state                                          | OPERATIONAL                | 'ALTER CLUSTER af5f3c1561e6c967c6337599461cf246 ADD `a_s58`'
[Tue Apr 30 05:20:39.409 2024] [312672] DEBUG: P01: syntax error, unexpected identifier near
cluster_c63021ddef800c5c809e4855b75bd80b_state_uuid          | 53fff817-06b1-11ef-92ee-eb1| 'ALTER CLUSTER d8783a3642045d36f66ec12ff41b402c ADD `a_s59`'
```

No clue why it says there is a syntax error when there is actually no error at all, but due to this, clusters are not created.

I can validate the logs on the server, and I did not remove anything.

Also, the waiting timeout occurs on the buddy just due to many logs and the debug build (so it's really really slow, so we are not in the allowed window of 30 seconds).

The best way to validate if there is any error:

```sql
select count(*) from sharding_queue where status = 'error';
+----------+
| count(*) |
+----------+
|       35 |
+----------+
```

**Comment by tomatolog** [comment]
that error is ok
```
Tue Apr 30 05:20:39.406 2024] [312651] DEBUG: P01: syntax error, unexpected identifier near
cluster_c_evs_state                                          | OPERATIONAL                | 'ALTER CLUSTER af5f3c1561e6c967c6337599461cf246 ADD `a_s58`'
```

it is the result of the hierarchical parser work lower parser produces error then up the hierarchy parser handle it or produce the next error and so on.

It also shows only in debug verbosity that is why it does not show in the regular work

**Comment by donhardman** [comment]
The issue isn't just about this error. We should try to reproduce it again because I lowered the timeout to 0.05 seconds in the buddy sources, and now I can see that sometimes clusters aren't being created (the same issue as before).

Please take a look again (buddy has modifications in your repo, and they should be kept).

**Comment by tomatolog** [comment]
I need buddy to log the error of the failed statement, now in the `searchd2.log` I see only the failed statement itself
```
[Fri May  3 13:46:37.822 2024] [463072] [BUDDY] [!] [127.0.0.1:31212] Queue query error: CREATE TABLE `a`
[Fri May  3 13:46:37.822 2024] [463072] [BUDDY] 			type='distributed' local='a_s1' local='a_s10' local='a_s11' local='a_s14' local='a_s15' local='a_s17' local='a_s19' local='a_s2' local='a_s21' local='a_s23' local='a_s24' local='a_s25' local='a_s26' local='a_s28' local='a_s31' local='a_s32' local='a_s34' local='a_s35' local='a_s37' local='a_s39' local='a_s4' local='a_s40' local='a_s41' local='a_s43' local='a_s46' local='a_s47' local='a_s48' local='a_s51' local='a_s52' local='a_s53' local='a_s55' local='a_s56' local='a_s58' local='a_s6' local='a_s7' local='a_s9' agent='127.0.0.1:31112:a_s0|127.0.0.1:31312:a_s0|127.0.0.1:31512:a_s0' agent='127.0.0.1:31112:a_s12|127.0.0.1:31312:a_s12|127.0.0.1:31412:a_s12' agent='127.0.0.1:31112:a_s13|127.0.0.1:31312:a_s13|127.0.0.1:31512:a_s13' agent='127.0.0.1:31112:a_s16|127.0.0.1:31312:a_s16|127.0.0.1:31412:a_s16' agent='127.0.0.1:31112:a_s20|127.0.0.1:31312:a_s20|127.0.0.1:31412:a_s20' agent='127.0.0.1:31112:a_s22|127.0.0.1:31312:a_s22|127.0.0.1:31512:a_s22' age
[Fri May  3 13:46:37.822 2024] [463072] [BUDDY] nt='127.0.0.1:31112:a_s29|127.0.0.1:31412:a_s29|127.0.0.1:31512:a_s29' agent='127.0.0.1:31112:a_s30|127.0.0.1:31412:a_s30|127.0.0.1:31512:a_s30' agent='127.0.0.1:31112:a_s38|127.0.0.1:31312:a_s38|127.0.0.1:31512:a_s38' agent='127.0.0.1:31112:a_s3|127.0.0.1:31412:a_s3|127.0.0.1:31512:a_s3' agent='127.0.0.1:31112:a_s42|127.0.0.1:31312:a_s42|127.0.0.1:31412:a_s42' agent='127.0.0.1:31112:a_s45|127.0.0.1:31412:a_s45|127.0.0.1:31512:a_s45' agent='127.0.0.1:31112:a_s49|127.0.0.1:31312:a_s49|127.0.0.1:31412:a_s49' agent='127.0.0.1:31112:a_s50|127.0.0.1:31312:a_s50|127.0.0.1:31512:a_s50' agent='127.0.0.1:31112:a_s54|127.0.0.1:31312:a_s54|127.0.0.1:31512:a_s54' agent='127.0.0.1:31112:a_s59|127.0.0.1:31312:a_s59|127.0.0.1:31412:a_s59' agent='127.0.0.1:31112:a_s5|127.0.0.1:31312:a_s5|127.0.0.1:31412:a_s5' agent='127.0.0.1:31112:a_s8|127.0.0.1:31312:a_s8|127.0.0.1:31412:a_s8' agent='127.0.0.1:31312:a_s18|127.0.0.1:31412:a_s18|127.0.0.1:31512:a_s18' agent='127.0.0
[Fri May  3 13:46:37.822 2024] [463072] [BUDDY] .1:31312:a_s27|127.0.0.1:31412:a_s27|127.0.0.1:31512:a_s27' agent='127.0.0.1:31312:a_s33|127.0.0.1:31412:a_s33|127.0.0.1:31512:a_s33' agent='127.0.0.1:31312:a_s36|127.0.0.1:31412:a_s36|127.0.0.1:31512:a_s36' agent='127.0.0.1:31312:a_s44|127.0.0.1:31412:a_s44|127.0.0.1:31512:a_s44' agent='127.0.0.1:31312:a_s57|127.0.0.1:31412:a_s57|127.0.0.1:31512:a_s57'
[Fri May  3 13:46:37.822 2024] [463072] [BUDDY] 		 
```

It is impossible to figure out what cause daemon to fail that statement and why buddy goes into repeat loop to recreate distributed table.

I figure out that previously in the `searchd2.log` there is a warning
```
[Fri May  3 13:46:37.819 2024] [463070] WARNING: table 'a': no such local table 'a_s39', SKIPPED
```
seems node2 joins the cluster while `a_s39` table was created and added into cluster at the node1 then the old issue that table that added into cluster while other node joins that cluster failed to sync into joining node. However without proper error from buddy it is very hard to figure it out.

**Comment by donhardman** [comment]
To show any errors that pop up when things are processing, we should run Buddy with the `--debugv` option.

Let's investigate what's happening with table `a_s39`. I can't find it on the server, while `a_s58` is there. It's normal for creating a distributed table to fail at first because we keep trying until it works while waiting for the other tables to show up.

Could you tell me where table `a_s39` is located? I can't find any record of it in the searchd2 log. Also, try running it with the `--debugv` switch so we can get a clearer picture of what's going on.

P.S. I've added a feature that gradually increases the timeout, but it hasn't made a significant difference based on my local tests. You can check it out here: https://github.com/manticoresoftware/manticoresearch-buddy/pull/282

**Comment by tomatolog** [comment]
should I use `main` branch of the buddy instead of the `sharding/inserts` to investigate the issue further?

**Comment by sanikolaev** [comment]
@tomatolog Discussed with Dmitrii. You should use `sharding/inserts`. Only in it there's timeout PT0.05S.

**Comment by tomatolog** [comment]
then I use these daemon options
```
        network_timeout = 120s
        client_timeout = 120s
        replication_connect_timeout = 30s
        replication_query_timeout = 120s
        replication_retry_count = 5
```
from the log files
```
--- query1.log
19:26:18.323 2024 conn 534 (127.0.0.1:58828) real 0.000 */ CREATE TABLE IF NOT EXISTS `a_s22` (id bigint) ;
19:26:18.326 2024 conn 534 (127.0.0.1:58828) real 0.002 */ /* DONE */;
19:26:18.328 2024 conn 537 (127.0.0.1:58848) real 0.000 */ ALTER CLUSTER ecb841b63a92fbe66097296d2d1c8375 ADD `a_s22`;
19:26:18.415 2024 conn 537 (127.0.0.1:58848) real 0.087 */ /* DONE */;
19:26:24.641 2024 conn 806 (127.0.0.1:42760) real 0.000 */ CREATE TABLE `a`
19:26:24.642 2024 conn 806 (127.0.0.1:42760) real 0.001 */ /* DONE */;
19:26:18.421 2024 conn 543 (127.0.0.1:58910) real 0.000 */ ALTER CLUSTER ecb841b63a92fbe66097296d2d1c8375 ADD `a_s23`;
19:26:18.482 2024 conn 543 (127.0.0.1:58910) real 0.060 */ /* DONE */;

--- query2.log
19:26:18.325 2024 conn 181 (127.0.0.1:44992) real 0.000 */ CREATE TABLE `a`
19:26:18.336 2024 conn 202 (127.0.0.1:45220) real 0.000 */ CREATE TABLE `a`
19:26:19.030 2024 conn 265 (127.0.0.1:45786) real 0.000 */ CREATE TABLE `a`
19:26:20.031 2024 conn 349 (127.0.0.1:46530) real 0.000 */ CREATE TABLE `a`
19:26:21.031 2024 conn 384 (127.0.0.1:46806) real 0.000 */ CREATE TABLE `a`
19:26:22.031 2024 conn 398 (127.0.0.1:46950) real 0.000 */ CREATE TABLE `a`
19:26:23.031 2024 conn 412 (127.0.0.1:42506) real 0.000 */ CREATE TABLE `a`
19:26:24.032 2024 conn 496 (127.0.0.1:43266) real 0.000 */ CREATE TABLE `a`
19:26:25.031 2024 conn 559 (127.0.0.1:43800) real 0.000 */ CREATE TABLE `a`
    
--- searchd1.log
19:26:24.647 2024] [1332733] [BUDDY] [127.0.0.1:31112] Queue query: CREATE TABLE `a`
19:26:39.996 2024] [1332709] [BUDDY] [6654de50ed3fb1.05111895] response data: {"version":2,"type":"sql response","message":{"error":"Waiting timeout exceeded."},"error_code":200} 

--- searchd2.log
19:26:18.325 2024] [1333343] DEBUG: P01: syntax error, unexpected CREATE near 'CREATE TABLE `a`
19:26:18.325 2024] [1333343] DEBUG: P01: syntax error, unexpected CREATE near 'CREATE TABLE `a`
19:26:18.325 2024] [1333343] WARNING: table 'a': no such local table 'a_s22', SKIPPED
19:26:18.336 2024] [1333346] WARNING: table 'a': no such local table 'a_s23', SKIPPED
19:26:18.337 2024] [1333326] [BUDDY] [127.0.0.1:31212] Queue query: CREATE TABLE `a`
19:26:18.337 2024] [1333326] [BUDDY] [!] [127.0.0.1:31212] Queue query error: CREATE TABLE `a`
19:26:18.337 2024] [1333326] [BUDDY] [127.0.0.1:31212] Queue query result [error]: {"error":"error adding table 'a': table 'a': no such local table 'a_s22'"} 
19:26:18.337 2024] [1333326] [BUDDY] [127.0.0.1:31212] Queue query result [error]: {"error":"error adding table 'a': table 'a': no such local table 'a_s23'"} 
19:26:19.030 2024] [1333329] WARNING: table 'a': no such local table 'a_s30', SKIPPED
19:26:19.036 2024] [1333347] [BUDDY] [127.0.0.1:31212] Queue query result [error]: {"error":"error adding table 'a': table 'a': no such local table 'a_s30'"} 
19:26:20.031 2024] [1333348] WARNING: table 'a': no such local table 'a_s40', SKIPPED
19:26:20.045 2024] [1333335] [BUDDY] [127.0.0.1:31212] Queue query result [error]: {"error":"error adding table 'a': table 'a': no such local table 'a_s40'"} 
19:26:21.032 2024] [1333327] WARNING: table 'a': no such local table 'a_s43', SKIPPED
19:26:21.036 2024] [1333343] [BUDDY] [127.0.0.1:31212] Queue query result [error]: {"error":"error adding table 'a': table 'a': no such local table 'a_s43'"} 
19:26:22.031 2024] [1333329] WARNING: table 'a': no such local table 'a_s43', SKIPPED
19:26:22.039 2024] [1333351] [BUDDY] [127.0.0.1:31212] Queue query result [error]: {"error":"error adding table 'a': table 'a': no such local table 'a_s43'"} 
19:26:23.032 2024] [1333340] WARNING: table 'a': no such local table 'a_s43', SKIPPED
19:26:23.041 2024] [1333348] [BUDDY] [127.0.0.1:31212] Queue query result [error]: {"error":"error adding table 'a': table 'a': no such local table 'a_s43'"} 
19:26:24.032 2024] [1333337] WARNING: table 'a': no such local table 'a_s53', SKIPPED
19:26:24.037 2024] [1333338] [BUDDY] [127.0.0.1:31212] Queue query result [error]: {"error":"error adding table 'a': table 'a': no such local table 'a_s53'"} 
```

I see that buddy at the `node2` creates distributed index `a` that use local index that was adding into cluster at the `node1`, ie buddy at the `node1` start adding `a_s23` at the `19:26:18.421` and finished operation at the `19:26:18.482` however at the prior to that buddy at the `node2` creates table `a` with the distributed index `a_s23` at the `19:26:18.325` and failed. Then next time it failed to creates table `a` with the distributed index `a_s30` and so on.

Some times buddy failed to wait for the `ALTER CLUSTER name ADD 'tableN' ` to finish as buddy from other nodes joins the cluster and alter that get into such join stuck waiting while all joins finish then only adds table into cluster.

I'd remove parallel processing of all the sharding steps from the buddy and make it only sequentially as for now it is impossible to investigate the case due to multiple event and that buddy does not report the final error but only `Waiting timeout exceeded` that is useless.
Or we could move sharding from the buddy into daemon that eliminate inconsistency between multiple buddies and multiple daemons.

**Comment by sanikolaev** [comment]
As discussed, reassigning back to you, @donhardman

**Comment by sanikolaev** [comment]
@tomatolog do you think it will help if we support `alter cluster C add tbl1, tbl2, tbl3` to lower the number of operations/transactions to be made inside Galera?

**Comment by tomatolog** [comment]
yes that could reduce the overall wait time as this will be the single transaction

**Comment by sanikolaev** [comment]
> yes that could reduce the overall wait time as this will be the single transaction

OK. Let's try this. 

Blocked by https://github.com/manticoresoftware/manticoresearch/issues/2308

**Comment by sanikolaev** [comment]
> Blocked by https://github.com/manticoresoftware/manticoresearch/issues/2308

Unblocked.

**Comment by donhardman** [comment]
I have implemented the logic on Buddy but discovered that the functionality isn't working as expected. Here's a minimal reproducible example (MRE):

```sql
mysql> show tables;
mysql> create cluster test;
mysql> create table a1;
mysql> create table a2;
mysql> alter cluster test add `a1`, `a2`;
ERROR 1064 (42000): unknown or wrong type of table ' `a2'
mysql>
mysql>
mysql> alter cluster test add `a1`, `a2`;
ERROR 1064 (42000): unknown or wrong type of table ' `a2'
mysql> show tables;
+-------+------+
| Index | Type |
+-------+------+
| a1    | rt   |
| a2    | rt   |
+-------+------+
mysql>
```

As we can see, it shows an error about the second table even though it was created in the same manner.

**Comment by donhardman** [comment]
As we discovered, we are missing support for parsing spaces between tables and enclosing the table in backticks (`).

**Comment by donhardman** [comment]
After extensive testing, I discovered that grouping multiple tables into a single request did not help. The daemon is still slow, replication is out of sync, and queries fail all the time when we try to create sharding.

We can continue to troubleshoot and implement fixes on the Buddy side, but I believe there's something wrong with the daemon. All queued queries from Buddy return no errors and execute properly, yet we still experience a 95% failure rate due to low PT timeout.

![Image](https://github.com/manticoresoftware/manticoresearch-buddy/assets/59067742/f21aa941-159e-4623-b5e8-7aef93143029)



![Image](https://github.com/manticoresoftware/manticoresearch-buddy/assets/59067742/21b328aa-bcfb-4585-ac11-d2b807458bcd)


P.S. All updates are pushed to the `sharding/inserts` branch of the buddy, and it's merged with the multiple queries fix and the `sharding/tests` branch in Manticoresearch merged with master.

P.P.S. **Reminder**: manticoresearch branch to use: `sharding/tests`

**Comment by donhardman** [comment]
I've discovered and fixed some critical issues on the Buddy side: https://github.com/manticoresoftware/manticoresearch-buddy/pull/317

However, the issue still persists, and the error "**replication connection failed: 7 'error in node state, must reinit**" remains active and should be addressed.

I've reduced the number of shards to 5 to make investigation easier. The problem is easily reproducible on 5 nodes with 3 RF, but not at all on 3 nodes, even with 2 or 3 RF.

I've pushed a simplified test to the same branch `sharding/inserts`. We can remove it later since it's only for investigation purposes and shouldn't be included with all the tests.

The code that I used to catch the issue:

```bash
counter=0;while true; do echo Runs: $((counter++));clt/clt test -d -t test/clt-tests/sharding/replication/mre.rec ghcr.io/manticoresoftware/manticoresearch:test-kit-latest; code=$(echo $?); echo Exit code: $code; if [ $code -eq 1 ]; then break; fi; done
```

![Image](https://github.com/manticoresoftware/manticoresearch-buddy/assets/59067742/b1065b18-49a8-4f8a-9467-1c7afd6e439f)

**Comment by tomatolog** [comment]
I see that buddy issue `CREATE TABLE `t` type='distributed' local='t_s1'` with the local table `t_s1` that comes from the cluster prior to joining the cluster, ie buddy joins cluster two times, 1st attempt failed then buddy issue 2nd attempt to join the cluster but then issue create table without waiting join to finish
```
[Sun Jul 14 17:26:55.323 2024] [676] [BUDDY] [127.0.0.1:49312] Queue query result [error]: {"error":"replication connection failed: 7 'error in node state, must reinit'"} 
[Sun Jul 14 17:26:55.323 2024] [676] [BUDDY] [!] [127.0.0.1:49312] Queue query error: JOIN CLUSTER ff01ef0d31440a59a0ad4b569caf8464 at '127.0.0.1:19312' 'ff01ef0d31440a59a0ad4b569caf8464' as path 
[Sun Jul 14 17:27:01.727 2024] [679] [BUDDY] [127.0.0.1:49312] Queue query result [error]: {"error":"error adding table 't': table 't': no such local table 't_s1'"} 
[Sun Jul 14 17:27:01.727 2024] [679] [BUDDY] [!] [127.0.0.1:49312] Queue query error: CREATE TABLE `t`
```

the general flow of the case at the `searchd1` table `t_s1` created and added into cluster well
```
/* Sun Jul 14 17:26:21.264 2024 conn 143 (127.0.0.1:43722) real 0.000 */ CREATE TABLE IF NOT EXISTS `t_s1` (id bigint) ;
[Sun Jul 14 17:26:21.264 2024] [527] DEBUG: P01: syntax error, unexpected CREATE near 'CREATE TABLE IF NOT EXISTS `t_s1` (id bigint) '
[Sun Jul 14 17:26:21.266 2024] [527] DEBUG: config saved, tables 5, clusters 3
[Sun Jul 14 17:26:21.266 2024] [505] RPL: committed seq 55
...
/* Sun Jul 14 17:26:21.550 2024 conn 157 (127.0.0.1:43866) real 0.000 */ ALTER CLUSTER ff01ef0d31440a59a0ad4b569caf8464 ADD t_s1;
[Sun Jul 14 17:26:21.550 2024] [497] DEBUG: P01: syntax error, unexpected identifier near 'ALTER CLUSTER ff01ef0d31440a59a0ad4b569caf8464 ADD t_s1'
[Sun Jul 14 17:26:21.550 2024] [497] RPL: node parse error: empty nodes list
[Sun Jul 14 17:26:21.550 2024] [497] RPL: alter 'ff01ef0d31440a59a0ad4b569caf8464' SST index 't_s1' to nodes 1: '127.0.0.1:19312'
[Sun Jul 14 17:26:21.550 2024] [497] RPL: alter 'ff01ef0d31440a59a0ad4b569caf8464' SST index 't_s1' to resolved nodes 0
```

then `searchd2` node joined cluster well and received that table
```
/* Sun Jul 14 17:26:21.929 2024 conn 69 (127.0.0.1:55550) real 0.000 */ JOIN CLUSTER ff01ef0d31440a59a0ad4b569caf8464 at '127.0.0.1:19312' 'ff01ef0d31440a59a0ad4b569caf8464' as path;
[Sun Jul 14 17:26:26.498 2024] [876] RPL: cluster 'ff01ef0d31440a59a0ad4b569caf8464' view nodes changed: 127.0.0.1:19312,127.0.0.1:19326:replication > 127.0.0.1:19312,127.0.0.1:19326:replication,127.0.0.1:29312,127.0.0.1:29324:replication
[Sun Jul 14 17:26:26.498 2024] [876] RPL: join 127.0.0.1:29312 to ff01ef0d31440a59a0ad4b569caf8464
...
[Sun Jul 14 17:26:26.535 2024] [557] RPL: reserve table 't_s1'
[Sun Jul 14 17:26:26.535 2024] [586] RPL: switching disk file /var/lib/manticore-2/ff01ef0d31440a59a0ad4b569caf8464/t_s1/t_s1.meta.new (-1>0), restart 0
[Sun Jul 14 17:26:26.535 2024] [584] RPL: switching disk file /var/lib/manticore-2/ff01ef0d31440a59a0ad4b569caf8464/t_s1/t_s1.ram.new (-1>1), restart 0
[Sun Jul 14 17:26:26.536 2024] [558] RPL: switching disk file /var/lib/manticore-2/ff01ef0d31440a59a0ad4b569caf8464/t_s1/t_s1.settings.new (-1>2), restart 0
[Sun Jul 14 17:26:26.536 2024] [561] RPL: remote cluster command 4(index_add_local), client 127.0.0.1:34854
[Sun Jul 14 17:26:26.536 2024] [561] RPL: rotating table 't_s1' content from /var/lib/manticore-2/ff01ef0d31440a59a0ad4b569caf8464/t_s1/t_s1
[Sun Jul 14 17:26:26.536 2024] [561] RPL: loading table 't_s1' into cluster 'ff01ef0d31440a59a0ad4b569caf8464' from /var/lib/manticore-2/ff01ef0d31440a59a0ad4b569caf8464/t_s1/t_s1
[Sun Jul 14 17:26:26.537 2024] [562] RPL: remote cluster command 5(synced), client 127.0.0.1:34864
[Sun Jul 14 17:26:26.537 2024] [562] RPL: join sync ff01ef0d31440a59a0ad4b569caf8464, UID 2fda1890-4206-11ef-bce4-a75c9a98fdd9:1, sent ok, tables 1, 
```

then `searchd4`  node is joining the cluster but failed due to low timeout
```
/* Sun Jul 14 17:26:22.205 2024 conn 76 (127.0.0.1:53720) real 0.000 */ JOIN CLUSTER ff01ef0d31440a59a0ad4b569caf8464 at '127.0.0.1:19312' 'ff01ef0d31440a59a0ad4b569caf8464' as path;
[Sun Jul 14 17:26:22.205 2024] [675] RPL: joining cluster 'ff01ef0d31440a59a0ad4b569caf8464', nodes: 127.0.0.1:19312
...
[Sun Jul 14 17:26:55.281 2024] [682] FATAL: /__w/manticoresearch/manticoresearch/build/galera-build/_deps/galera_populate-src/gcs/src/gcs.cpp:gcs_open():1514: Failed to open channel 'ff01ef0d31440a59a0ad4b569caf8464' at 'gcomm://127.0.0.1:19326': -110 (Connection timed out)
[Sun Jul 14 17:26:55.281 2024] [682] FATAL: /__w/manticoresearch/manticoresearch/build/galera-build/_deps/galera_populate-src/galera/src/replicator_smm.cpp:connect():356: gcs connect failed: Connection timed out
[Sun Jul 14 17:26:55.281 2024] [682] RPL: cluster 'ff01ef0d31440a59a0ad4b569caf8464' wait to finish
[Sun Jul 14 17:26:55.281 2024] [682] RPL: deleting provider of cluster ff01ef0d31440a59a0ad4b569caf8464
[Sun Jul 14 17:26:55.320 2024] [682] RPL: cluster 'ff01ef0d31440a59a0ad4b569caf8464' finished, cluster deleted
```
then buddy again issue join at the node `searchd4`
```
/* Sun Jul 14 17:27:01.728 2024 conn 213 (127.0.0.1:50614) real 0.000 */ JOIN CLUSTER ff01ef0d31440a59a0ad4b569caf8464 at '127.0.0.1:19312' 'ff01ef0d31440a59a0ad4b569caf8464' as path;
[Sun Jul 14 17:27:01.728 2024] [689] RPL: joining cluster 'ff01ef0d31440a59a0ad4b569caf8464', nodes: 127.0.0.1:19312
...

[Sun Jul 14 17:27:01.722 2024] [691] RPL: committed seq 202
/* Sun Jul 14 17:27:01.722 2024 conn 198 (127.0.0.1:50488) real 0.000 */ CREATE TABLE `t` type='distributed' local='t_s1' local='t_s2' local='t_s3' agent='127.0.0.1:19312:t_s0|127.0.0.1:39312:t_s0|127.0.0.1:59312:t_s0' agent='127.0.0.1:29312:t_s4|127.0.0.1:39312:t_s4|127.0.0.1:59312:t_s4'
[Sun Jul 14 17:27:01.722 2024] [695] DEBUG: P01: syntax error, unexpected CREATE near 'CREATE TABLE `t` type='distributed' local='t_s1' local='t_s2' local='t_s3' agent='127.0.0.1:19312:t_s0|127.0.0.1:39312:t_s0|127.0.0.1:59312:t_s0' agent='127.0.0.1:29312:t_s4|127.0.0.1:39312:t_s4|127.0.0.1:59312:t_s4'
[Sun Jul 14 17:27:01.722 2024] [695] WARNING: table 't': no such local table 't_s1', SKIPPED
[Sun Jul 14 17:27:01.723 2024] [684] RPL: committed seq 203
...
[Sun Jul 14 17:27:03.298 2024] [688] RPL: loading table 't_s1' into cluster 'ff01ef0d31440a59a0ad4b569caf8464' from /var/lib/manticore-4/ff01ef0d31440a59a0ad4b569caf8464/t_s1/t_s1
[Sun Jul 14 17:27:03.299 2024] [686] RPL: join sync ff01ef0d31440a59a0ad4b569caf8464, UID 2fda1890-4206-11ef-bce4-a75c9a98fdd9:1, sent ok, tables 1, 
```

but at the same time buddy issue create table with the table that will come from the cluster and `CREATE TABLE` failed. The join process at the node `searchd4` finished well and table got received well but buddy just hang then return general error to client `Waiting timeout exceeded`

For me it is still not clear the buddy commands flow at the different nodes and why it can not returns the last error to client but reply with `Waiting timeout exceeded`

I see no issues with the daemon work here

**Comment by donhardman** [comment]
The fix for the Buddy side was implemented in this pull request: https://github.com/manticoresoftware/manticoresearch-buddy/pull/321

I've also cherry-picked it to the `sharding/inserts` branch that we use for running CLT tests.

I think now we can try to track down the cluster initialization error that actually reproduced on my setup on the first try:

```
––– output –––
+ /var/log/manticore-3/searchd.log:[Wed Jul 17 10:46:24.928 2024] [544] [BUDDY] [127.0.0.1:39312] Queue query result [error]: {"error":"replication connection failed: 7 'error in node state, must reinit'"}
```

**Comment by sanikolaev** [comment]
Pls review https://manticoresearch.slack.com/archives/D04043PE9JQ/p1722235150946239

**Comment by sanikolaev** [comment]
Here's how I test it on `dev2` in `/home/snikolaev/repl/1`:

```
snikolaev@dev2:~/repl/1$ for m in `seq 1 100`; do echo "attempt $m"; for n in 1 2; do echo -----------------${n}---------------; kill -9 `cat ~/repl/$n/searchd.pid`; sleep 3; rm -fr ~/repl/$n/data/*; mkdir ~/repl/$n/data/c; rm ~/repl/$n/*.log; searchd -c ~/repl/$n/manticore${n}.conf --logreplication; done; sleep 3; mysql -P10301 -h0 -v -e "create cluster c"; mysql -P20301 -h0 -v -e "join cluster c at '127.0.0.1:10201'"; time mysql -P10301 -h0 -v -e "create table c:t(id bigint) shards=2 rf=2 timeout=60" || break; done; mysql -P10301 -h0 -e "select * from _sharding_queue; select * from _sharding_state; select * from _sharding_table"
```

Buddy is checked out in `/home/snikolaev/repl/buddy`. 
The 2nd node used in this case is located here `/home/snikolaev/repl/2`.

**Comment by donhardman** [comment]
I introduced some new fixes and found that 240 runs without issues for now.

I pushed it to the same `sharding/inserts` branch and also created a pull request to merge into the main: https://github.com/manticoresoftware/manticoresearch-buddy/pull/334

I think there are no issues on Buddy now, except for the possible "must reinit."

**Comment by sanikolaev** [comment]
As discussed, the `must reinit` can be still reproduced. @donhardman pls remind, how this can be reproduced: only with CLT or without it, on what server etc.

**Comment by donhardman** [comment]
So as of now, the buddy issue seems to be fixed @tomatolog. Let's proceed with researching the real reason for the "must reinit" that happens, especially more often in case we lower the Galera timeout to PT0.1S and lower.

We should take the daemon from the branch `tests/sharding`. I rebased it to the latest master. To catch the issue, we can use the following code from the manticoresearch directory. We should make sure that we have also checked out the `clt` repository to the same directory.

```
date; counter=0; while true; do echo Runs: $((counter++)); clt/clt test -d -t test/clt-tests/sharding/replication/mre.rec ghcr.io/manticoresoftware/manticoresearch:test-kit-latest; code=$?; echo Exit code: $code; if [ $code -eq 1 ]; then break; fi; done; date
```

The test that fails is extracted and simplified in mre.rec.

I saw that Stas has already made some changes, but to catch it with all following logs, we should try to increase `timeout=120` in case it's impossible to reproduce.

I tried to run this test and got a failure on the first try, so it should now be easy to debug since the Buddy issue has been eliminated.

![Image](https://github.com/user-attachments/assets/7fbb33e2-1e3e-443e-b73d-e21c484248fa)

**Comment by tomatolog** [comment]
the last clt test that I run with daemon from the `tests/sharding` branch  like
```
date; counter=0; while true; do echo Runs: $((counter++)); clt/clt test -d -t test/clt-tests/sharding/replication/mre.rec ghcr.io/manticoresoftware/manticoresearch:test-kit-latest; code=$?; echo Exit code: $code; if [ $code -eq 1 ]; then break; fi; done; date
```
end the same as it failed previously at https://github.com/manticoresoftware/manticoresearch-buddy/issues/255#issuecomment-2227449151

daemon at the node 3 creates distributed index prior that node joins the cluster and receives the index from the cluster by join, ie the flow

here is a view from the node 2 - at the node 2 table `t_s2` created then added into cluster then node 3 joins the cluster and expect table `t_s2` to sync into node 3
```
--- s2
/* Fri Aug 16 15:34:36.708 2024 conn 142 (127.0.0.1:46510) real 0.000 */ CREATE TABLE IF NOT EXISTS `t_s2` (id bigint) ;
[Fri Aug 16 15:34:36.708 2024] [833430] DEBUG: P01: syntax error, unexpected CREATE near 'CREATE TABLE IF NOT EXISTS `t_s2` (id bigint) '
[Fri Aug 16 15:34:36.709 2024] [833430] DEBUG: config saved, tables 5, clusters 3
/* Fri Aug 16 15:34:36.709 2024 conn 142 (127.0.0.1:46510) real 0.001 */ /* DONE */;
/* Fri Aug 16 15:34:36.711 2024 conn 145 (127.0.0.1:46536) real 0.000 */ ALTER CLUSTER de7f4b3f5b479de3e61e93e6e9673788 ADD t_s2;
[Fri Aug 16 15:34:36.711 2024] [833409] DEBUG: P01: syntax error, unexpected identifier near 'ALTER CLUSTER de7f4b3f5b479de3e61e93e6e9673788 ADD t_s2'
[Fri Aug 16 15:34:36.711 2024] [833409] RPL: node parse error: empty nodes list
[Fri Aug 16 15:34:36.711 2024] [833409] RPL: alter 'de7f4b3f5b479de3e61e93e6e9673788' SST index 't_s2' to nodes 1: '127.0.0.1:31212'
[Fri Aug 16 15:34:36.711 2024] [833409] RPL: alter 'de7f4b3f5b479de3e61e93e6e9673788' SST index 't_s2' to resolved nodes 0
[Fri Aug 16 15:34:36.711 2024] [833409] RPL: node parse error: empty nodes list
[Fri Aug 16 15:34:36.711 2024] [833409] RPL: alter 'de7f4b3f5b479de3e61e93e6e9673788' adding index 't_s2'
[Fri Aug 16 15:34:36.748 2024] [833409] RPL: alter 'de7f4b3f5b479de3e61e93e6e9673788' added index 't_s2'
/* Fri Aug 16 15:34:36.748 2024 conn 145 (127.0.0.1:46536) real 0.037 */ /* DONE */;
[Fri Aug 16 15:34:36.760 2024] [833415] [BUDDY] [37811µs] manticore request: query=ALTER+CLUSTER+de7f4b3f5b479de3e61e93e6e9673788+ADD+t_s2 
[Fri Aug 16 15:34:36.760 2024] [833415] [BUDDY] [127.0.0.1:31212] Queue query result [processed]: [{"total":0,"error":"","warning":""}] 
...
[Fri Aug 16 15:34:41.298 2024] [833743] RPL: cluster 'de7f4b3f5b479de3e61e93e6e9673788' view nodes changed: 127.0.0.1:31212,127.0.0.1:31232:replication > 127.0.0.1:31212,127.0.0.1:31232:replication,127.0.0.1:31412,127.0.0.1:31430:replication
[Fri Aug 16 15:34:41.402 2024] [833743] RPL: donate de7f4b3f5b479de3e61e93e6e9673788 to 127.0.0.1:31412, gtid 0b4849da-5be5-11ef-a125-0b84d805978b:1, bypass 0
[Fri Aug 16 15:34:41.402 2024] [833743] RPL: sending cluster 'de7f4b3f5b479de3e61e93e6e9673788' indexes 1 't_s2'...'t_s2'
[Fri Aug 16 15:34:41.464 2024] [833743] RPL: synced cluster de7f4b3f5b479de3e61e93e6e9673788
```

here is a view from the node 3 - node 3 joins the cluster to node 2 the joins failed due to low Galera timeout to PT0.1S but the node 3 skip that failure then proceed to creating distributed table that it thinks node 3 received from the successful join 
```
/* Fri Aug 16 15:34:37.651 2024 conn 340 (127.0.0.1:60244) real 0.000 */ JOIN CLUSTER de7f4b3f5b479de3e61e93e6e9673788 at '127.0.0.1:31212' 'de7f4b3f5b479de3e61e93e6e9673788' as path;
[Fri Aug 16 15:34:37.651 2024] [833471] DEBUG: P01: syntax error, unexpected JOIN near 'JOIN CLUSTER de7f4b3f5b479de3e61e93e6e9673788 at '127.0.0.1:31212' 'de7f4b3f5b479de3e61e93e6e9673788' as path'
[Fri Aug 16 15:34:37.651 2024] [833471] RPL: joining cluster 'de7f4b3f5b479de3e61e93e6e9673788', nodes: 127.0.0.1:31212
...
[Fri Aug 16 15:35:10.736 2024] [833475] FATAL: /__w/manticoresearch/manticoresearch/build/galera-build/_deps/galera_populate-src/gcs/src/gcs.cpp:gcs_open():1514: Failed to open channel 'de7f4b3f5b479de3e61e93e6e9673788' at 'gcomm://127.0.0.1:31232': -110 (Connection timed out)
/* Fri Aug 16 15:35:10.850 2024 conn 340 (127.0.0.1:60244) real 33.198 */ /* DONE */;
...
/* Fri Aug 16 15:35:10.852 2024 conn 472 (127.0.0.1:44082) real 0.000 */ CREATE TABLE `t`
			type='distributed' local='t_s0' local='t_s2' agent='127.0.0.1:31112:t_s1|127.0.0.1:31212:t_s1|127.0.0.1:31412:t_s1';
[Fri Aug 16 15:35:10.852 2024] [833483] DEBUG: P01: syntax error, unexpected CREATE near 'CREATE TABLE `t`
			type='distributed' local='t_s0' local='t_s2' agent='127.0.0.1:31112:t_s1|127.0.0.1:31212:t_s1|127.0.0.1:31412:t_s1'
[Fri Aug 16 15:35:10.852 2024] [833483] WARNING: table 't': no such local table 't_s2', SKIPPED
/* Fri Aug 16 15:35:10.852 2024 conn 472 (127.0.0.1:44082) real 0.000 */ /* DONE */;
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] [33199011µs] manticore request: query=JOIN+CLUSTER+de7f4b3f5b479de3e61e93e6e9673788+at+%27127.0.0.1%3A31212%27+%27de7f4b3f5b479de3e61e93e6e9673788%27+as+path 
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] [127.0.0.1:31312] Queue query result [error]: {"error":"replication connection failed: 7 'error in node state, must reinit'"} 
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] [!] [127.0.0.1:31312] Queue query error: JOIN CLUSTER de7f4b3f5b479de3e61e93e6e9673788 at '127.0.0.1:31212' 'de7f4b3f5b479de3e61e93e6e9673788' as path 
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] [127.0.0.1:31312] Queue query: CREATE TABLE `t`
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] 			type='distributed' local='t_s0' local='t_s2' agent='127.0.0.1:31112:t_s1|127.0.0.1:31212:t_s1|127.0.0.1:31412:t_s1'
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] 		 
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] [375µs] manticore request: query=CREATE+TABLE+%60t%60%0A%09%09%09type%3D%27distributed%27+local%3D%27t_s0%27+local%3D%27t_s2%27+agent%3D%27127.0.0.1%3A31112%3At_s1%7C127.0.0.1%3A31212%3At_s1%7C127.0.0.1%3A31412%3At_s1%27%0A%09%09 
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] [127.0.0.1:31312] Queue query result [error]: {"error":"error adding table 't': table 't': no such local table 't_s2'"} 
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] [!] [127.0.0.1:31312] Queue query error: CREATE TABLE `t`
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] 			type='distributed' local='t_s0' local='t_s2' agent='127.0.0.1:31112:t_s1|127.0.0.1:31212:t_s1|127.0.0.1:31412:t_s1'
[Fri Aug 16 15:35:10.857 2024] [833462] [BUDDY] 		 
```

I see that nothing changes - buddy does not count the current state but switching states one after another like all works ok. I also see the buddy lost the errors and does not report them to client. That is why it is impossible to figure out what goes wrong without long debug sessions with local reproducible case

**Comment by tomatolog** [comment]
I use Codership Galera from the galera-3 package to test vanila Galera work in this case and see the same behavior. I use the 
[Dockerfile.zip](https://github.com/user-attachments/files/16783574/Dockerfile.zip) to build local Docker image on top of the `ghcr.io/manticoresoftware/manticoresearch:test-kit-latest` image and added `export GALERA_SONAME=/usr/lib/galera/libgalera_smm.so` to all `start-searchd-X.recb` to allow daemon to use that Galera. It works well however the case is still here.

I use clt from the https://github.com/manticoresoftware/manticoresearch/tree/tests/sharding and buddy from the https://github.com/manticoresoftware/manticoresearch-buddy/tree/sharding/inserts . At the buddy branch I made fixes for join cluster to use the same Galera options as the create cluster use. And also set new Galera options to set timeout to 10 second in case join waits the donor node.
But buddy still does not count the current state but switching states one after another like all works ok, ie if I run
```
date; counter=0; while true; do echo Runs: $((counter++)); sudo chown -R stas:stas ./clt/*; rm clt/bin/x86_64/logs/?/*.log; clt/clt test -d -t test/clt-tests/sharding/replication/mre.rec tmg-clt; code=$?; echo Exit code: $code; if [ $code -eq 1 ]; then break; fi; done; date
```

clt finishes with the error on 2-4 loop and the `searchd.log`  at the node 4 I see
```
[Wed Aug 28 12:42:15.723 2024] [692] RPL: joining cluster 'a917c70802428e5e82d5df875d568e6f', nodes: 127.0.0.1:29312
...
[Wed Aug 28 12:42:21.336 2024] [695] FATAL: /__w/manticoresearch/manticoresearch/build/galera-build/_deps/galera_populate-src/gcs/src/gcs.cpp:gcs_open():1514: Failed to open channel 'a917c70802428e5e82d5df875d568e6f' at 'gcomm://127.0.0.1:29326': -110 (Connection timed out)
...
[Wed Aug 28 12:42:21.372 2024] [710] DEBUG: P01: syntax error, unexpected CREATE near 'CREATE TABLE `t`
			type='distributed' local='t_s1' local='t_s2' local='t_s3' agent='127.0.0.1:19312:t_s0|127.0.0.1:39312:t_s0|127.0.0.1:59312:t_s0' agent='127.0.0.1:29312:t_s4|127.0.0.1:39312:t_s4|127.0.0.1:59312:t_s4'
		'
[Wed Aug 28 12:42:21.372 2024] [710] WARNING: table 't': no such local table 't_s2', SKIPPED
[Wed Aug 28 12:42:21.373 2024] [713] [BUDDY] [5647868µs] manticore request: query=JOIN+CLUSTER+a917c70802428e5e82d5df875d568e6f+at+%27127.0.0.1%3A29312%27+%27a917c70802428e5e82d5df875d568e6f%27+as+path%2C+%27gmcast.peer_timeout%3DPT3S%3Bevs.install_timeout%3DPT5S%3Bevs.delayed_keep_period%3DPT10S%3Bpc.wait_prim_timeout%3DPT5S%27+as+options 
[Wed Aug 28 12:42:21.373 2024] [713] [BUDDY] [127.0.0.1:49312] Queue query result [error]: {"error":"replication connection failed: 7 'error in node state, must reinit'"} 
[Wed Aug 28 12:42:21.373 2024] [713] [BUDDY] [!] [127.0.0.1:49312] Queue query error: JOIN CLUSTER a917c70802428e5e82d5df875d568e6f at '127.0.0.1:29312' 'a917c70802428e5e82d5df875d568e6f' as path, 'gmcast.peer_timeout=PT3S;evs.install_timeout=PT5S;evs.delayed_keep_period=PT10S;pc.wait_prim_timeout=PT5S' as options 
[Wed Aug 28 12:42:21.373 2024] [713] [BUDDY] [127.0.0.1:49312] Queue query: CREATE TABLE `t`
[Wed Aug 28 12:42:21.373 2024] [713] [BUDDY] 			type='distributed' local='t_s1' local='t_s2' local='t_s3' agent='127.0.0.1:19312:t_s0|127.0.0.1:39312:t_s0|127.0.0.1:59312:t_s0' agent='127.0.0.1:29312:t_s4|127.0.0.1:39312:t_s4|127.0.0.1:59312:t_s4'
```

- join the cluster `a917c70802428e5e82d5df875d568e6f` started at `12:42:15.723`
- join failed at `12:42:21.336` after 6 seconds
- buddy does not wait for the join result but start creation of distributed table `t` with local table `t_s2` this node should receive after join the cluster. This started at `12:42:21.372`
- buddy reports the join failure at `12:42:21.373`
- budy reports the create table failure at `12:42:21.373`
After that there is no any retry to join the cluster `a917c70802428e5e82d5df875d568e6f` or create distributed table `t`


here is similar reports from the clt runs with the vanila Galera

failure at the node 5
```
--- s5
[Wed Aug 28 11:50:11.710 2024] [786] RPL: joining cluster 'a057ba664134027b5a09f5b29d8bc9aa', nodes: 127.0.0.1:29312
...
[Wed Aug 28 11:50:17.366 2024] [789] FATAL: ./gcs/src/gcs.cpp:gcs_open():1475: Failed to open channel 'a057ba664134027b5a09f5b29d8bc9aa' at 'gcomm://127.0.0.1:29328': -110 (Connection timed out)
...
[Wed Aug 28 11:50:17.393 2024] [791] DEBUG: P01: syntax error, unexpected CREATE near 'CREATE TABLE `t`
			type='distributed' local='t_s0' local='t_s3' local='t_s4' agent='127.0.0.1:19312:t_s1|127.0.0.1:29312:t_s1|127.0.0.1:49312:t_s1' agent='127.0.0.1:29312:t_s2|127.0.0.1:39312:t_s2|127.0.0.1:49312:t_s2'
...
[Wed Aug 28 11:50:17.393 2024] [791] WARNING: table 't': no such local table 't_s4', SKIPPED
[Wed Aug 28 11:50:17.402 2024] [788] [BUDDY] [5680881µs] manticore request: query=JOIN+CLUSTER+a057ba664134027b5a09f5b29d8bc9aa+at+%27127.0.0.1%3A29312%27+%27a057ba664134027b5a09f5b29d8bc9aa%27+as+path%2C+%27gmcast.peer_timeout%3DPT3S%3Bevs.install_timeout%3DPT5S%3Bevs.delayed_keep_period%3DPT10S%3Bpc.wait_prim_timeout%3DPT5S%27+as+options 
[Wed Aug 28 11:50:17.402 2024] [788] [BUDDY] [127.0.0.1:59312] Queue query result [error]: {"error":"replication connection failed: 7 'error in node state, must reinit'"} 
[Wed Aug 28 11:50:17.402 2024] [788] [BUDDY] [!] [127.0.0.1:59312] Queue query error: JOIN CLUSTER a057ba664134027b5a09f5b29d8bc9aa at '127.0.0.1:29312' 'a057ba664134027b5a09f5b29d8bc9aa' as path, 'gmcast.peer_timeout=PT3S;evs.install_timeout=PT5S;evs.delayed_keep_period=PT10S;pc.wait_prim_timeout=PT5S' as options 
[Wed Aug 28 11:50:17.402 2024] [788] [BUDDY] [127.0.0.1:59312] Queue query: CREATE TABLE `t`
[Wed Aug 28 11:50:17.402 2024] [788] [BUDDY] 			type='distributed' local='t_s0' local='t_s3' local='t_s4' agent='127.0.0.1:19312:t_s1|127.0.0.1:29312:t_s1|127.0.0.1:49312:t_s1' agent='127.0.0.1:29312:t_s2|127.0.0.1:39312:t_s2|127.0.0.1:49312:t_s2'
```

failure at the node 3
```
---s3
[Wed Aug 28 07:47:48.404 2024] [635] RPL: joining cluster 'a057ba664134027b5a09f5b29d8bc9aa', nodes: 127.0.0.1:29312
[Wed Aug 28 07:47:55.658 2024] [633] FATAL: ./gcs/src/gcs.cpp:gcs_open():1475: Failed to open channel 'a057ba664134027b5a09f5b29d8bc9aa' at 'gcomm://127.0.0.1:29328': -110 (Connection timed out)

[Wed Aug 28 07:47:55.685 2024] [659] DEBUG: P01: syntax error, unexpected CREATE near 'CREATE TABLE `t`
			type='distributed' local='t_s0' local='t_s2' local='t_s4' agent='127.0.0.1:19312:t_s1|127.0.0.1:29312:t_s1|127.0.0.1:49312:t_s1' agent='127.0.0.1:19312:t_s3|127.0.0.1:49312:t_s3|127.0.0.1:59312:t_s3'
		'
[Wed Aug 28 07:47:55.686 2024] [659] WARNING: table 't': no such local table 't_s4', SKIPPED
[Wed Aug 28 07:47:55.689 2024] [658] [BUDDY] [!] [127.0.0.1:39312] Queue query error: CREATE TABLE `t`
[Wed Aug 28 07:47:55.689 2024] [658] [BUDDY] 			type='distributed' local='t_s0' local='t_s2' local='t_s4' agent='127.0.0.1:19312:t_s1|127.0.0.1:29312:t_s1|127.0.0.1:49312:t_s1' agent='127.0.0.1:19312:t_s3|127.0.0.1:49312:t_s3|127.0.0.1:59312:t_s3'
[Wed Aug 28 07:47:55.689 2024] [658] [BUDDY] [!] [127.0.0.1:39312] Queue query error: JOIN CLUSTER a057ba664134027b5a09f5b29d8bc9aa at '127.0.0.1:29312' 'a057ba664134027b5a09f5b29d8bc9aa' as path, 'gmcast.peer_timeout=PT0.1S;evs.install_timeout=PT5S;evs.delayed_keep_period=PT10S;pc.wait_prim_timeout=PT5S' as options 
```

I fixed the buddy logic to use the same Galera option for create cluster and join the cluster and found Galera options that set small join timeout and allow to fail earlie. However I can not figure out the Buddy logic and the flow and can not find the point there table should wait for the join to finish or the join loop retries on failure

**Comment by donhardman** [comment]
I have implemented a fix to adjust the timing for long-lasting queries, such as 'join cluster'. I also resolved a minor issue where queries would continue executing despite errors. However, the problem persists. It seems impossible to recover when this occurs, even with a reinitialisation.

Could you please validate these changes with your tests to see if the issue with Buddy has been resolved?

**Comment by tomatolog** [comment]
I modify clt from the https://github.com/manticoresoftware/manticoresearch/tree/tests/sharding ecea8fe76552532c0219d1029b0ea0b7393126b0 and see that sharding works ok now, ie cluster could fails to join but the retry works ok and after attempts sharded index got created. 
I checked a loop
```
date; counter=0; while true; do echo Runs: $((counter++)); sudo chown -R stas:stas ./clt/*; rm clt/bin/x86_64/logs/?/*.log; clt/clt test -d -t test/clt-tests/sharding/replication/mre.rec tmg-clt; code=$?; echo Exit code: $code; if [ $code -eq 1 ]; then break; fi; done; date
```

for 100 iterations and see that sharding clt test does not produce bad exit code

**Comment by donhardman** [comment]
I can confirm that we have no issues.

I suggest moving parameters as defaults to the Manticore Search code instead of keeping them in Buddy.

```
gmcast.peer_timeout=PT3S;evs.install_timeout=PT5S;evs.delayed_keep_period=PT10S;pc.wait_prim_timeout=PT5S
```

**Comment by donhardman** [comment]
Got error on run #145

```
+ *************************** 1. row ***************************
+ id: 56574764374321159
+ node: 127.0.0.1:39312
+ query: JOIN CLUSTER a917c70802428e5e82d5df875d568e6f at '127.0.0.1:29312' 'a917c70802428e5e82d5df875d568e6f' as path, 'gmcast.peer_timeout=PT3S;evs.install_timeout=PT5S;evs.delayed_keep_period=PT10S;pc.wait_prim_timeout=PT5S' as options
+ wait_for_id: 56574764374031585
+ tries: 10
+ status: error
+ created_at: 1726110189521
+ updated_at: 1726110205232
+ duration: 19
```

But if nothing to do with it, anyway it's better now.

**Comment by sanikolaev** [comment]
> I'd make these daemon options defaults related to sharding then if all work as usual for most of users move these into replication defaults.

I agree. @donhardman let's do it.

**Comment by donhardman** [comment]
I have moved parameters into this pull request: https://github.com/manticoresoftware/manticoresearch-buddy/pull/356

**Comment by donhardman** [comment]
Closing, as pull request was merged
